src/elm/ComponentHelper.elm

```diff
@@ -53,19 +72,33 @@ update msg model =
 
 subscriptions : Model -> Sub Msg
 subscriptions model =
-    parentMessage decodeMessage
+    Sub.batch
+        [ componentIn componentMessageDecoder
+        , coordinatorIn CoordinatorMessage
+        ]
+
+
+componentMessageDecoder : Decode.Value -> Msg
+componentMessageDecoder value =
+    case
+        Decode.decodeValue
+            (Decode.map ComponentMessage ComponentMsg.decoder)
+            value
+    of
+        Ok msg ->
+            msg
+
+        Err err ->
+            Unknown err
+
+
+port coordinatorIn : (Decode.Value -> msg) -> Sub msg
+
+
+port coordinatorOut : Decode.Value -> Cmd msg
 
 
-decodeMessage : Decode.Value -> Msg
-decodeMessage value =
-    Decode.decodeValue
-        (Decode.field "data"
-            (Decode.oneOf
-                [ Decode.map Navigate Navigate.decoder ]
-            )
-        )
-        value
-        |> Result.withDefault (Unknown value)
+port componentIn : (Decode.Value -> msg) -> Sub msg
 
 
-port parentMessage : (Decode.Value -> msg) -> Sub msg
+port componentOut : Decode.Value -> Cmd msg
```

Comment:
I'm starting to settle on a structure for message passing where there's a pair of input/output ports for talking to each independent app. So the component helper has a pair of ports for sending and receiving messages with the component app, and a pair of ports for sending and receiving messages with the coordinator (parent window)